src/Microsoft.AspNetCore.Routing/Matchers/HttpMethodMatcherPolicy.cs

```diff
@@ -47,38 +54,94 @@ public IReadOnlyList<PolicyNodeEdge> GetEdges(IReadOnlyList<Endpoint> endpoints)
         {
             // The algorithm here is designed to be preserve the order of the endpoints
             // while also being relatively simple. Preserving order is important.
-            var allHttpMethods = endpoints
-                    .SelectMany(e => GetHttpMethods(e))
-                    .Distinct()
-                    .OrderBy(m => m); // Sort for testability
 
-            var dictionary = new Dictionary<string, List<Endpoint>>();
-            foreach (var httpMethod in allHttpMethods)
+            // First, build a dictionary of all possible HTTP method/CORS combinations
```

Comment:
I rewrote this to make it pretty verbose, but way less spooky. It should be relatively comprehensible by now.